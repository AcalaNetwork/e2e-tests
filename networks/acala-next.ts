import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acalaNext' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karuraNext' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala_runtime.wasm',
        kusama: './wasm/karura_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
