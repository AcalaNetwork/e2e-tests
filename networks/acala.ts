import { Config } from './types'

export type Vars = {
  relayToken: string
  relayLiquidToken: string
}

export default {
  polkadot: {
    name: 'acala' as const,
    endpoint: 'wss://acala-rpc-0.aca-api.network',
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
  },
  kusama: {
    name: 'karura' as const,
    endpoint: 'wss://karura-rpc-0.aca-api.network',
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
  },
  config: ({ alice, relayToken, relayLiquidToken }) => ({
    storages: {
      System: {
        account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        accounts: [
          [[alice.address, { Token: relayToken }], { free: 10 * 1e12 }],
          [[alice.address, { Token: relayLiquidToken }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings', 'toBondPool'],
        // so that bump era won't trigger unbond
        relayChainCurrentEra: '0x64000000',
      },
      EvmAccounts: {
        accounts: [[['0x82a258cb20e2adb4788153cd5eb5839615ece9a0'], alice.address]],
        evmAddresses: [[[alice.address], '0x82a258cb20e2adb4788153cd5eb5839615ece9a0']],
      },
      PolkadotXcm: {
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const acala = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  dot: { Token: 'DOT' },
  dai: { Erc20: '0x54a37a01cd75b616d63e0ab665bffdb0143c52ae' },
  wbtc: { ForeignAsset: 5 },
} as const

export const karura = {
  paraId: 2000,
  paraAccount: '13YMK2eYoAvStnzReuxBjMrAvPXmmdsURwZvc62PrdXimbNy',
  ksm: { Token: 'KSM' },
  usdt: { ForeignAsset: 7 },
  dai: { Erc20: '0x4bb6afb5fa2b07a5d1c499e1c3ddb5a15e709a71' },
} as const
