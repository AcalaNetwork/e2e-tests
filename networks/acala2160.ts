import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2160' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2160' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2160-dev.wasm',
        kusama: './wasm/karura-2160-dev.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
