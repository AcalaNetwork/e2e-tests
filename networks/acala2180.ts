import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2180' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2180' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2180.wasm',
        kusama: './wasm/karura-2180.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
