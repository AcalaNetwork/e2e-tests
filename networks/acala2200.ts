import { Config } from './types'

import acalaConfig, { Vars } from './acala'

export default {
  ...acalaConfig,
  polkadot: {
    ...acalaConfig.polkadot,
    name: 'acala2200' as const,
  },
  kusama: {
    ...acalaConfig.kusama,
    name: 'karura2200' as const,
  },
  config: (opt) => ({
    ...acalaConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/acala-2200.wasm',
        kusama: './wasm/karura-2200.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
