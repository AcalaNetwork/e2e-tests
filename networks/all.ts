import { Config } from './types'

import acala2160Config from './acala2160'
import acalaConfig from './acala'
import hydraDXConfig from './hydraDX'
import moonbeamConfig from './moonbeam'
import polkadot9381Config from './polkadot9381'
import polkadotConfig from './polkadot'
import statemintConfig from './statemint'

const all = {
  polkadot: polkadotConfig,
  polkadot9381: polkadot9381Config,
  statemint: statemintConfig,
  acala: acalaConfig,
  acala2160: acala2160Config,
  moonbeam: moonbeamConfig,
  hydraDX: hydraDXConfig,
} satisfies Record<string, Config>

export default all
