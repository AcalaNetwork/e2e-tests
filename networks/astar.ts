import { Config } from './types'

export type Vars = {
  relayToken: string
}

export default {
  polkadot: {
    name: 'astar' as const,
    endpoint: 'wss://astar.api.onfinality.io/public-ws',
    relayToken: '340282366920938463463374607431768211455',
  },
  kusama: {
    name: 'shiden' as const,
    endpoint: 'wss://shiden.api.onfinality.io/public-ws',
    relayToken: '340282366920938463463374607431768211455',
  },
  config: ({ alice, relayToken }) => ({
    storages: {
      System: {
        account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Assets: {
        account: [[[relayToken, alice.address], { balance: 10 * 1e12 }]],
      },
      Sudo: {
        key: alice.address,
      },
      PolkadotXcm: {
        // avoid sending xcm version change notifications to makes things faster
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const astar = {
  paraId: 2006,
  paraAccount: '13YMK2eZzuFY1WZGagpYtTgbWBWGdoUD2CtrPj1mQPjY8Ldc',
  dot: 340282366920938463463374607431768211455n,
  astr: { Concrete: { parents: 0, interior: 'Here' } },
  aca: 18446744073709551616n,
  usdt: 4294969280n,
} as const

export const shiden = {
  paraId: 2007,
  paraAccount: 'F7fq1jNy74AqkJ1DP4KqSrWtnTGtXfNVoDwFhTvvPxUvJaq',
  ksm: 340282366920938463463374607431768211455n,
  sdn: { Concrete: { parents: 0, interior: 'Here' } },
  kar: 18446744073709551618n,
  usdt: 4294969280n,
} as const
