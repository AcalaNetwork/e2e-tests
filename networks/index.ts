import { SetupOption, setupContext } from '@acala-network/chopsticks-testing'
import { connectParachains, connectVertical } from '@acala-network/chopsticks'
import dotenv from 'dotenv'

import { testingPairs } from '../helpers'

import { Config, Context, NetworkKind } from './types'
import networkDefs from './all'

dotenv.config()

const toNumber = (value: string | undefined): number | undefined => {
  if (value === undefined) {
    return undefined
  }

  return Number(value)
}

export type Network = Awaited<ReturnType<typeof setupContext>> & {
  options: SetupOption
  config: (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]
}
export type NetworkNames = (typeof networkDefs)[keyof typeof networkDefs][NetworkKind]['name']

export const networkCreator = {} as Record<
  NetworkNames,
  (options?: Partial<SetupOption>) => (ctx: Context) => Promise<Network>
>

const relaychains = ['polkadot', 'kusama'] as const

for (const def of Object.values(networkDefs)) {
  for (const relaychain of relaychains) {
    const config = def[relaychain]
    const { endpoint, name } = config
    const upperName = name.toUpperCase()
    networkCreator[name] = (options?: Partial<SetupOption>) => async (ctx: Context) => {
      const setupConfig = (def as Config).config({
        network: relaychain,
        ...config,
        ...ctx,
      })

      const finalOptions: SetupOption = {
        timeout: 600000,
        wasmOverride: process.env[`${upperName}_WASM`],
        blockNumber: toNumber(process.env[`${upperName}_BLOCK_NUMBER`]),
        endpoint: process.env[`${upperName}_ENDPOINT`] ?? endpoint,
        db: process.env.DB_PATH,
        ...setupConfig.options,
        ...options,
      }

      const network = await setupContext(finalOptions)

      await network.dev.setStorage(setupConfig.storages)

      return {
        ...network,
        config,
        options: finalOptions,
      }
    }
  }
}

export const createContext = () => testingPairs()

export const createNetworks = async (
  networkOptions: Partial<Record<NetworkNames, Partial<SetupOption> | undefined>>,
  context = createContext()
) => {
  const ret = {} as Record<NetworkNames, Network>

  let wasmOverriden = false

  for (const [name, options] of Object.entries(networkOptions) as [NetworkNames, Partial<SetupOption> | undefined][]) {
    ret[name] = await networkCreator[name](options)(context)
    wasmOverriden ||= !!ret[name].options.wasmOverride
  }

  const relaychainName = Object.keys(ret).filter(
    (x) => x.startsWith('polkadot') || x.startsWith('kusama')
  )[0] as NetworkKind
  const { [relaychainName]: relaychain, ...parachains } = ret

  if (relaychain) {
    for (const parachain of Object.values(parachains)) {
      await connectVertical(relaychain.chain, parachain.chain)
    }
  }

  const parachainList = Object.values(parachains).map((i) => i.chain)
  if (parachainList.length > 0) {
    await connectParachains(parachainList)
  }

  if (wasmOverriden) {
    // trigger runtime upgrade if needed (due to wasm override)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
    // handle xcm version message if needed (due to wasm override triggered xcm version upgrade)
    for (const chain of Object.values(ret)) {
      await chain.dev.newBlock()
    }
  }

  return ret
}

// to be compatible with old code

const networks = {} as Record<NetworkNames, (options?: Partial<SetupOption>) => Promise<Network>>

for (const [name, creator] of Object.entries(networkCreator)) {
  networks[name as NetworkNames] = async (options?: Partial<SetupOption>) => creator(options)(createContext())
}

export default networks
