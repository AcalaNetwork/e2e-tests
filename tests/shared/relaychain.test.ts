import { afterAll, beforeEach, describe, it } from 'vitest'
import { connectVertical } from '@acala-network/chopsticks'

import { check, checkEvents, checkSystemEvents, checkUmp, sendTransaction, testingPairs } from '../helper'
import networks from '../networks'

describe.each([
  {
    para: 'acala',
    relay: 'polkadot',
    relayToken: 'DOT',
    relayLiquidToken: 'LDOT',
  },
  {
    para: 'karura',
    relay: 'kusama',
    relayToken: 'KSM',
    relayLiquidToken: 'LKSM',
  },
])('$para <-> $relay', async ({ para, relay, relayToken, relayLiquidToken }) => {
  const relaychain = await networks[relay as keyof typeof networks]()
  const parachain = await networks[para as keyof typeof networks]()
  await connectVertical(relaychain.chain, parachain.chain)

  const { alice } = testingPairs()

  afterAll(async () => {
    await relaychain.teardown()
    await parachain.teardown()
  })

  beforeEach(async () => {
    await parachain.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      Tokens: {
        Accounts: [
          [[alice.address, { Token: relayToken }], { free: 10 * 1e12 }],
          [[alice.address, { Token: relayLiquidToken }], { free: 100 * 1e12 }],
        ],
      },
      Sudo: {
        Key: alice.address,
      },
      Homa: {
        // avoid impact test outcome
        $removePrefix: ['redeemRequests', 'unbondings'],
      },
    })
    await relaychain.dev.setStorage({
      System: {
        Account: [[[alice.address], { data: { free: 10 * 1e12 } }]],
      },
      ParasDisputes: {
        // those can makes block building super slow
        $removePrefix: ['disputes'],
      },
    })
  })

  it('parachain transfer assets to relaychain', async () => {
    const tx = await sendTransaction(
      parachain.api.tx.xTokens
        .transfer(
          {
            Token: relayToken,
          },
          1e12,
          {
            V1: {
              parents: 1,
              interior: {
                X1: {
                  AccountId32: {
                    network: 'Any',
                    id: alice.addressRaw,
                  },
                },
              },
            },
          },
          'Unlimited'
        )
        .signAsync(alice)
    )

    await parachain.chain.newBlock()

    await check(relaychain.api.query.system.account(alice.address)).toMatchSnapshot()

    await checkEvents(tx, 'xTokens').toMatchSnapshot()
    await checkUmp(parachain).toMatchSnapshot()

    await relaychain.chain.newBlock()

    await check(parachain.api.query.tokens.accounts(alice.address, { Token: relayToken })).toMatchSnapshot()
    await checkSystemEvents(relaychain, 'ump').toMatchSnapshot()
  })

  it('Homa stake works', async () => {
    const tx1 = await sendTransaction(parachain.api.tx.homa.mint(1e12).signAsync(alice, { nonce: 0 }))
    const tx2 = await sendTransaction(
      parachain.api.tx.sudo.sudo(parachain.api.tx.homa.forceBumpCurrentEra(0)).signAsync(alice, { nonce: 1 })
    )

    await parachain.chain.newBlock()

    await checkEvents(tx1, 'homa').redact({ number: 1 }).toMatchSnapshot()
    await checkEvents(tx2, { section: 'homa', method: 'CurrentEraBumped' }).toMatchSnapshot()
    await checkUmp(parachain).redact({ number: true, hex: true }).toMatchSnapshot()

    await relaychain.chain.newBlock()

    await checkSystemEvents(relaychain, 'ump', 'staking').redact({ address: true, number: true }).toMatchSnapshot()
  })

  it('Homa redeem unbond works', async () => {
    const tx3 = await sendTransaction(
      parachain.api.tx.homa.requestRedeem(10 * 1e12, false).signAsync(alice, { nonce: 0 })
    )
    const tx4 = await sendTransaction(
      parachain.api.tx.sudo.sudo(parachain.api.tx.homa.forceBumpCurrentEra(0)).signAsync(alice, { nonce: 1 })
    )

    await parachain.chain.newBlock()

    await checkEvents(tx3, 'homa').toMatchSnapshot()
    await checkEvents(tx4, 'homa').toMatchSnapshot()

    await relaychain.chain.newBlock()

    await checkSystemEvents(relaychain, 'ump', 'staking').redact({ address: true }).toMatchSnapshot()
  })
})
