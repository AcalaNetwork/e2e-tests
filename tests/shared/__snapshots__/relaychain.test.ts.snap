// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`'acala' <-> 'polkadot' > Homa redeem unbond works > events 1`] = `
[
  {
    "data": {
      "allowFastMatch": false,
      "liquidAmount": 10000000000000,
      "redeemer": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "RequestedRedeem",
    "section": "homa",
  },
]
`;

exports[`'acala' <-> 'polkadot' > Homa redeem unbond works > events 2`] = `
[
  {
    "data": {
      "newEraIndex": "(rounded 1000)",
    },
    "method": "CurrentEraBumped",
    "section": "homa",
  },
  {
    "data": {
      "eraIndexWhenUnbond": "(rounded 1000)",
      "liquidAmount": 10000000000000,
      "redeemer": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
      "unbondingStakingAmount": "(rounded 1200000000000)",
    },
    "method": "RedeemedByUnbond",
    "section": "homa",
  },
]
`;

exports[`'acala' <-> 'polkadot' > Homa redeem unbond works > system events 1`] = `
[
  {
    "data": {
      "amount": "(rounded 1200000000000)",
      "stash": "(address)",
    },
    "method": "Unbonded",
    "section": "staking",
  },
  {
    "data": [
      "(hash)",
      {
        "Complete": "(rounded 11000000000)",
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
]
`;

exports[`'acala' <-> 'polkadot' > Homa stake works > events 1`] = `
[
  {
    "data": {
      "liquidAmountAddedToVoid": "(rounded 3000000000)",
      "liquidAmountReceived": "(rounded 8000000000000)",
      "minter": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
      "stakingCurrencyAmount": 1000000000000,
    },
    "method": "Minted",
    "section": "homa",
  },
]
`;

exports[`'acala' <-> 'polkadot' > Homa stake works > events 2`] = `
[
  {
    "data": {
      "newEraIndex": "(rounded 1000)",
    },
    "method": "CurrentEraBumped",
    "section": "homa",
  },
]
`;

exports[`'acala' <-> 'polkadot' > Homa stake works > system events 1`] = `
[
  {
    "data": [
      "(hash)",
      {
        "Complete": "(number)",
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
]
`;

exports[`'acala' <-> 'polkadot' > Homa stake works > ump 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
          "weightLimit": {
            "limited": "(number)",
          },
        },
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "accountId32": {
                  "id": "(hex)",
                  "network": {
                    "any": null,
                  },
                },
              },
            },
            "parents": "(number)",
          },
          "maxAssets": "(number)",
        },
      },
    ],
  },
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
        ],
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "transact": {
          "call": {
            "encoded": "(hex)",
          },
          "originType": "SovereignAccount",
          "requireWeightAtMost": "(number)",
        },
      },
      {
        "refundSurplus": null,
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "parachain": "(number)",
              },
            },
            "parents": "(number)",
          },
          "maxAssets": "(number)",
        },
      },
    ],
  },
]
`;

exports[`'acala' <-> 'polkadot' > parachain transfer assets to relaychain > events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000,
          },
          "id": {
            "Concrete": {
              "interior": "Here",
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X1": {
            "AccountId32": {
              "id": "(hash)",
              "network": "Any",
            },
          },
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 1,
          },
        },
      },
      "sender": "23y3WetbNi6rDMgHmyRDjgpb7PnhgPotuPPawxruTMLYTLzG",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`'acala' <-> 'polkadot' > parachain transfer assets to relaychain > system events 1`] = `
[
  {
    "data": [
      "(hash)",
      {
        "Complete": 4000000000,
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
]
`;

exports[`'acala' <-> 'polkadot' > parachain transfer assets to relaychain > ump 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "accountId32": {
                  "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                  "network": {
                    "any": null,
                  },
                },
              },
            },
            "parents": 0,
          },
          "maxAssets": 1,
        },
      },
    ],
  },
]
`;

exports[`'acala' <-> 'polkadot' > parachain transfer assets to relaychain 1`] = `
{
  "consumers": 0,
  "data": {
    "feeFrozen": 0,
    "free": 10000000000000,
    "miscFrozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 0,
  "sufficients": 0,
}
`;

exports[`'acala' <-> 'polkadot' > parachain transfer assets to relaychain 2`] = `
{
  "free": 9000000000000,
  "frozen": 0,
  "reserved": 0,
}
`;

exports[`'karura' <-> 'kusama' > Homa redeem unbond works > events 1`] = `
[
  {
    "data": {
      "allowFastMatch": false,
      "liquidAmount": 10000000000000,
      "redeemer": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "RequestedRedeem",
    "section": "homa",
  },
]
`;

exports[`'karura' <-> 'kusama' > Homa redeem unbond works > events 2`] = `
[
  {
    "data": {
      "newEraIndex": "(rounded 4900)",
    },
    "method": "CurrentEraBumped",
    "section": "homa",
  },
  {
    "data": {
      "eraIndexWhenUnbond": "(rounded 4900)",
      "liquidAmount": 10000000000000,
      "redeemer": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
      "unbondingStakingAmount": "(rounded 1300000000000)",
    },
    "method": "RedeemedByUnbond",
    "section": "homa",
  },
]
`;

exports[`'karura' <-> 'kusama' > Homa redeem unbond works > system events 1`] = `
[
  {
    "data": {
      "amount": "(rounded 1300000000000)",
      "stash": "(address)",
    },
    "method": "Unbonded",
    "section": "staking",
  },
  {
    "data": [
      "(hash)",
      {
        "Complete": "(rounded 15000000000)",
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
]
`;

exports[`'karura' <-> 'kusama' > Homa stake works > events 1`] = `
[
  {
    "data": {
      "liquidAmountAddedToVoid": "(rounded 600000000)",
      "liquidAmountReceived": "(rounded 8000000000000)",
      "minter": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
      "stakingCurrencyAmount": 1000000000000,
    },
    "method": "Minted",
    "section": "homa",
  },
]
`;

exports[`'karura' <-> 'kusama' > Homa stake works > events 2`] = `
[
  {
    "data": {
      "newEraIndex": "(rounded 4900)",
    },
    "method": "CurrentEraBumped",
    "section": "homa",
  },
]
`;

exports[`'karura' <-> 'kusama' > Homa stake works > system events 1`] = `
[
  {
    "data": [
      "(hash)",
      {
        "Complete": "(number)",
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
  {
    "data": {
      "amount": "(number)",
      "stash": "(address)",
    },
    "method": "Bonded",
    "section": "staking",
  },
  {
    "data": [
      "(hash)",
      {
        "Complete": "(number)",
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
]
`;

exports[`'karura' <-> 'kusama' > Homa stake works > ump 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
          "weightLimit": {
            "limited": "(number)",
          },
        },
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "accountId32": {
                  "id": "(hex)",
                  "network": {
                    "any": null,
                  },
                },
              },
            },
            "parents": "(number)",
          },
          "maxAssets": "(number)",
        },
      },
    ],
  },
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
        ],
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": "(number)",
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": "(number)",
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "transact": {
          "call": {
            "encoded": "(hex)",
          },
          "originType": "SovereignAccount",
          "requireWeightAtMost": "(number)",
        },
      },
      {
        "refundSurplus": null,
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "parachain": "(number)",
              },
            },
            "parents": "(number)",
          },
          "maxAssets": "(number)",
        },
      },
    ],
  },
]
`;

exports[`'karura' <-> 'kusama' > parachain transfer assets to relaychain > events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000,
          },
          "id": {
            "Concrete": {
              "interior": "Here",
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X1": {
            "AccountId32": {
              "id": "(hash)",
              "network": "Any",
            },
          },
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 1,
          },
        },
      },
      "sender": "rPizfonc8MPuEsMxufAytHjpRW7a2YrXWHKibaJXe9ZtjKx",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`'karura' <-> 'kusama' > parachain transfer assets to relaychain > system events 1`] = `
[
  {
    "data": [
      "(hash)",
      {
        "Complete": "(rounded 300000000)",
      },
    ],
    "method": "ExecutedUpward",
    "section": "ump",
  },
]
`;

exports[`'karura' <-> 'kusama' > parachain transfer assets to relaychain > ump 1`] = `
[
  {
    "v2": [
      {
        "withdrawAsset": [
          {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
        ],
      },
      {
        "clearOrigin": null,
      },
      {
        "buyExecution": {
          "fees": {
            "fun": {
              "fungible": 1000000000000,
            },
            "id": {
              "concrete": {
                "interior": {
                  "here": null,
                },
                "parents": 0,
              },
            },
          },
          "weightLimit": {
            "unlimited": null,
          },
        },
      },
      {
        "depositAsset": {
          "assets": {
            "wild": {
              "all": null,
            },
          },
          "beneficiary": {
            "interior": {
              "x1": {
                "accountId32": {
                  "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                  "network": {
                    "any": null,
                  },
                },
              },
            },
            "parents": 0,
          },
          "maxAssets": 1,
        },
      },
    ],
  },
]
`;

exports[`'karura' <-> 'kusama' > parachain transfer assets to relaychain 1`] = `
{
  "consumers": 0,
  "data": {
    "feeFrozen": 0,
    "free": 10000000000000,
    "miscFrozen": 0,
    "reserved": 0,
  },
  "nonce": 0,
  "providers": 0,
  "sufficients": 0,
}
`;

exports[`'karura' <-> 'kusama' > parachain transfer assets to relaychain 2`] = `
{
  "free": 9000000000000,
  "frozen": 0,
  "reserved": 0,
}
`;
